### PR TITLE
test(dependency): Improve test coverage

### PR DESCRIPTION
**Pull request by ts1994tw**
## Summary
This pull request enhances the test suite for Jotai's dependency management, particularly focusing on complex scenarios. The following improvement has been made:

## New Test Cases
1. **Complex Dependency Chains**
   - Introduced a test case that verifies Jotai's ability to handle and update complex chains of derived atoms, including async derivations.

## Coverage Improvements
- **Multi-level Derivations**: Verifies that updates properly propagate through multiple levels of derived atoms, including async operations.

## Check List
- [x] pnpm run prettier for formatting code and docs

Appreciated for having me your time on reviewing changes.

### CODE REVIEW

File: tests/vanilla/dependency.test.tsx [hunk 1]
```diff
@@ -274,3 +274,43 @@ it('refreshes deps for each async read', async () => {
...
+
+it('handles errors in async atoms', async () => {
+  const errorAtom = atom(async () => {
+    throw new Error('Test error')
+  })
+
+  const store = createStore()
+
+  await expect(store.get(errorAtom)).rejects.toThrow('Test error')
+})
```
Comment (dai-shi) [inline tests/vanilla/dependency.test.tsx:286]: This doesn't seem a dependency test. Maybe move to some other file?

File: tests/vanilla/dependency.test.tsx [hunk 2]
```diff
@@ -274,3 +274,43 @@ it('refreshes deps for each async read', async () => {
...
+
+it('handles complex dependency chains', async () => {
+  const baseAtom = atom(1)
+  const derived1 = atom((get) => get(baseAtom) * 2)
+  const derived2 = atom((get) => get(derived1) + 1)
+  const asyncDerived = atom(async (get) => {
+    const value = get(derived2)
+    await new Promise((resolve) => setTimeout(resolve, 10))
```
Comment (dai-shi) [inline tests/vanilla/dependency.test.tsx:294]: We don't use timers in the test.
Please do something like this:
https://github.com/pmndrs/jotai/blob/9bd8ed7b0e523af5366f6b762104241a0bb71203/tests/vanilla/dependency.test.tsx#L11

File: tests/vanilla/dependency.test.tsx [hunk 3]
```diff
@@ -274,3 +274,43 @@ it('refreshes deps for each async read', async () => {
...
+  store.set(baseAtom, 2)
+  expect(await store.get(asyncDerived)).toBe(10)
+})
+
+it('handles large number of atoms efficiently', () => {
+  const store = createStore()
+  const atoms = new Array(1000).fill(null).map((_, i) => atom(i))
+  const derivedAtom = atom((get) => atoms.reduce((sum, a) => sum + get(a), 0))
+
+  const start = performance.now()
```
Comment (dai-shi) [inline tests/vanilla/dependency.test.tsx:310]: Hmm, I don't think we do performance tests in our unit tests.
Can you try something with `./benchmarks`? It's not a test that runs on CI, but you can try more complex cases and share the result in a discussion.

**Review by dai-shi — CHANGES_REQUESTED** [review]
Thanks for working on this.
Here are some comments.

File: tests/vanilla/dependency.test.tsx (hunk 3 above)
Comment (backbone87) [reply tests/vanilla/dependency.test.tsx:310]: I think the existing benchmark suite also needs some refinement. I did some resource tracking a while back and memory consumption was not lining up with what I thought the tests are testing for

Comment (dai-shi) [reply tests/vanilla/dependency.test.tsx:310]: > I think the existing benchmark suite also needs some refinement.

I actually agree. Not just refinement, you can create a brand new benchmark suite. We can even delete existing one, if yours covers it. That would be pretty helpful.

File: tests/vanilla/dependency.test.tsx (hunk 2 above)
Comment (ts1994tw) [reply tests/vanilla/dependency.test.tsx]: @dai-shi Appreciated for indication and this below is adjustment which follows comment here.
[6186a57](https://github.com/pmndrs/jotai/pull/2759/commits/6186a5731564688e3cd2c2e317e67b5902d9ac0c)

File: tests/vanilla/dependency.test.tsx (hunk 3 above)
Comment (ts1994tw) [reply tests/vanilla/dependency.test.tsx]: @backbone87 Appreciated for sharing detections you noticed before. @dai-shi I prefer to leave this topic aside and it should independently handle by another clear and goal-focused pull request. These changes have been removed at this moment. 🙇

File: tests/vanilla/dependency.test.tsx (hunk 1 above)
Comment (ts1994tw) [reply tests/vanilla/dependency.test.tsx]: @dai-shi Appreciated for comments. You're absolutely right, it should be moved out of here. These changes have been removed from this pull request at this moment, after that, this pull request will be goal-focused enough and for dependency test only.

**Review by dai-shi — APPROVED** [review]
Thanks for your contribution.

**Pull request merged**